File: FAANG Interview Solutions/Set_Matrix_Zeroes.py
'''
Given an m x n integer matrix matrix, if an element is 0, set its entire row and column to 0's, and return the matrix.

You must do it in place.

Input: matrix = [[1,1,1],[1,0,1],[1,1,1]]
Output: [[1,0,1],[0,0,0],[1,0,1]]

Constraints:

m == matrix.length
n == matrix[0].length
1 <= m, n <= 200
-231 <= matrix[i][j] <= 231 - 1
'''
#better
class Solution(object):
    def setZeroes(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: void Do not return anything, modify matrix in-place instead.
        """
        R = len(matrix)
        C = len(matrix[0])
        rows, cols = set(), set()

        # Essentially, we mark the rows and columns that are to be made zero
        for i in range(R):
            for j in range(C):
                if matrix[i][j] == 0:
                    rows.add(i)
                    cols.add(j)

        # Iterate over the array once again and using the rows and cols sets, update the elements
        for i in range(R):
            for j in range(C):
                if i in rows or j in cols:
                    matrix[i][j] = 0
                    
#alternate
class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        m,n = len(matrix), len(matrix[0])
        row, col = [1]*m, [1]*n
        
        
        for r in range(m):
            for c in range(n):
                if matrix[r][c] == 0:
                    row[r] = 0
                    col[c] = 0
                    
        for r in range(m):
            if row[r] == 0:
                matrix[r] = [0]*n
                
        for c in range(n):
            if col[c] == 0:
                for i in range(m):
                    matrix[i][c] = 0
                    
        return matrix
      
      
 
        

File: Striver SDE Sheet #180/Day 1/Set_Matrix_Zeroes.java
/*  
Given an m x n integer matrix matrix, if an element is 0, set its entire row and column to 0's.
You must do it in place.

Input: matrix = [[1,1,1],[1,0,1],[1,1,1]]
Output: [[1,0,1],[0,0,0],[1,0,1]]

Constraints:
m == matrix.length
n == matrix[0].length
1 <= m, n <= 200
-231 <= matrix[i][j] <= 231 - 1

*/

class Solution {
    public void setZeroes(int[][] matrix) {
        boolean isCol = false;
        int rows = matrix.length;
        int columns = matrix[0].length;
        
         //STEP 1
        //1st traverse the matrix linearly
        for(int i=0; i< rows; i++){
            //if element at ith row and 1st col is 0, set isCol = true
            if(matrix[i][0] == 0){
                isCol = true;
            }
        
            
            //STEP 2
            for(int j=1; j<columns;j++){
                //if elm at ith row and jth col is 0
                if(matrix[i][j] == 0){
                    //set leftmost elm on that row to 0
                    matrix[i][0] = 0;
                    //set topmost elm on that col to 0
                    matrix[0][j] = 0;
                }
            }
            
        }
            //STEP 3
            //traversing the matrix backwards since dummy R and C needs to be reserved
            for(int i = rows-1; i>=0; i--){
                for(int j = columns-1; j>=1; j--){
                    //if leftmost or topmost is 0, update elm in ith row & jth col to 0
                    if(matrix[i][0] ==0 || matrix[0][j] == 0){
                        matrix[i][j] = 0;
                    }
                }
                
            //STEP 4 -  now we can update dummies
            //so, check if any elemnt in the 1st column is 0 i.e. isCol == true
            //then set col[0] to 0
                if (isCol)
                    matrix[i][0] = 0;
            }
               
        }
        
    }
